### DISCUSSION TITLE
Atypical (Maybe?) Custom Workflow for IPAM

### DISCUSSION THREAD

**Discussion by JeffHochberg** [netbox-community/netbox#10734] [Q&A]
Hello,

I have a bit of a unique requirement. I was thinking Netbox would be ideal for this... I am open to suggestions.

Keep in mind that what I'm about to cover is currently being done with a spreadsheet - I'm sure you can imagine how prone this is to human error.

### Questions

As you review this - please think about:

- Is it possible to build out this sort of workflow in Netbox?
- If so, how much effort would you anticipate is required?
- Is there another application that you can think of that would work better?

### Requirement

For each end-user/subscriber provision the following (in a single transaction):

1x /31 IP from Prefix A

1x /32 IP address from Prefix B

1x /32 IP address from Prefix C

### Workflow

The idea is to the following:

#### Netbox Admin

1. Select 3x prefixes for non-admins to allocate addresses from
2. Log into Netbox
3. Add them to Prefixes/Subnets (or if there's a better place)

#### Task-Oriented Worker

Then, the task-oriented worker would:

1. Log into Netbox
2. Enter username
3. Provision 3x addresses (as mentioned earlier - 1x /31 from Prefix A, 1x /32 from Prefix B, and 1x /32 from Prefix C)

Each IP should be the next available in the prefix to ensure there isn't any overlap and address space is used efficiently

### Technical Specifics

Imagine 3 separate prefixes (fictitious RFC 1918 addresses used for this example):

1. Prefix 1: 10.10.10.0 (1x /31 per customer - next available)
2. Prefix 2: 10.20.20.0 (1x /32 per customer - next available)
3. Prefix 3: 10.30.30.0 (1x /32 per customer - next available)

### Example

Let's say someone connects to Netbox to provision an end-user - I would like to use Netbox to assist with the following:

| End-User | Prefix 1 (/31 - 2 IPs) | Prefix 2 (/32 - 1 IP) | Prefix 3 (/32 - 1 IP) |
| -------- | ---------------------- | --------------------- | --------------------- |
| User A   | 10.10.10.**0/31**      | 10.20.20.1/32         | 10.30.30.1/32         |
| User B   | 10.10.10.**2/31**      | 10.20.20.2/32         | 10.30.30.2/32         |
| User C   | 10.10.10.**4/31**      | 10.20.20.3/32         | 10.30.30.3/32         |
| User D   | 10.10.10.**6/31**      | 10.20.20.4/32         | 10.30.30.4/32         |

Thank you for taking the time to review this! I look forward to reviewing your thoughts.

Best Regards,

-JeffH

**Comment by Kludge-As-A-Service** [comment]
Seems like a custom script could do the job, I see no major roadblock. Netbox admin would set a specific role (or tag) to the 3 container prefixes, then the custom provisioning script would have a reference to this role/tag and could handle the per-user prefix reservations. Assuming all prefixes within a container are the same size (which is the case, from what you say), Netbox seems to be able to handle the new prefix calculation by itself (at least there is a button for that in the UI so I assume this is available in the code as well) so you don't have to code that (not that it would be too difficult anyway).

And you can link users and their prefixes either with : 
- the built-in Tenancy feature
- or add a custom field to prefixes, which would reference a "contact" object.

**Reply by jeremystretch** [reply]
Agreed; you could very easily write a [custom script](https://docs.netbox.dev/en/stable/customization/custom-scripts/) to accomplish this. You'll have to put some thought into the provisioning logic, but the Prefix model's [`get_available_ips()`](https://github.com/netbox-community/netbox/blob/4c504870e029058d8e1cfeaaea84ac31b93a55e9/netbox/ipam/models/ip.py#L507) method will do a lot of the lifting for you.

**Comment by jeffh-cloudflare** [comment]
Thanks for the feedback guys! I really appreciate it.
